zkevm-circuits/src/mpt_circuit/branch.rs

```diff
@@ -1,307 +1,338 @@
-                            // `mod_node_hash_rlc` the same for all branch children
-                            let mod_node_hash_rlc = ctx.accumulators.mod_node_rlc(is_s);
-                            require!(a!(mod_node_hash_rlc) => a!(mod_node_hash_rlc, -1));
-                            // `contains_placeholder_leaf` the same for all branch children
-                            require!(contains_placeholder_leaf(meta, ctx.clone(), is_s, 0)
-                                => contains_placeholder_leaf(meta, ctx.clone(), is_s, -1));
-                        }
-                    }}
+                    // Extension node RLC
+                    let ext_node_rlc = (config.ext_rlp_key.rlc(&r), config.ext_mult.expr()).rlc_chain(
+                        config.ext_rlp_value[is_s.idx()].rlc_rlp(&mut cb.base, &r)
+                    );
+                    // Branch value data zero check
+                    cb.set_length_c(config.ext_rlp_value[is_s.idx()].num_bytes());
 
-                    // If we have a branch child, we can only have branch child or branch init in the previous row.
-                    require!(or::expr([is_branch_child.prev(), is_branch_init.prev()]) => true);
-                    // When `node_index` != 0
-                    ifx!{node_index => {
-                        // `node_index` increases by 1 for each branch child.
-                        require!(node_index => node_index.prev() + 1.expr());
-                    }}
+                    // The branch expected in the extension node
+                    ext_branch_rlp_rlc[is_s.idx()] = config.ext_rlp_value[is_s.idx()].rlc_rlp(&mut cb.base, &r);
 
-                    // We need to ensure that the only change in `S` and `C` proof occurs
-                    // at `modified_index` so that only a single change can be done.
-                    // We check `s_main.rlp = c_main.rlp` everywhere except at `modified_index`.
-                    // (except rlp1, rlp1 is used to keep track of number of bytes processed).
-                    let not_at_modification = node_index.expr() - modified_index.expr();
-                    ifx!{not_at_modification => {
-                        for (s_byte, c_byte) in s_main.rlp_bytes().iter().skip(1)
-                            .zip(c_main.rlp_bytes().iter().skip(1))
-                        {
-                            require!(a!(s_byte) => a!(c_byte));
+                    // Check if the extension node is in its parent.
+                    let (rlc, num_bytes, is_not_hashed) = {
+                        if is_s {
+                            config.ext_is_not_hashed = LtGadget::construct(&mut cb.base, config.ext_rlp_key.num_bytes(), HASH_WIDTH.expr());
                         }
+                        (ext_node_rlc.expr(), config.ext_rlp_key.num_bytes(), config.ext_is_not_hashed.expr())
+                    };
+                    // TODO(Brecht): why not if it's a placeholder?
```

Comment:
Placeholder branch just fills the rows (mostly, except for the `drifted_node` RLC) to preserve the parallel layout, so the hash that is needed to be checked in the proof with the placeholder branch is the hash of a leaf to be in the branch above the placeholder branch.